src/car_backed_blockstore.rs

```diff
@@ -0,0 +1,272 @@
+                reader.seek(SeekFrom::Start(*offset))?;
+                let cid = Cid::read_bytes(&mut *reader)?;
+                assert_eq!(cid, *k);
+                let cid_len = reader.stream_position()? - *offset;
+                let data_len = *length - usize::try_from(cid_len).unwrap();
+                let mut data = vec![0; data_len];
+                reader.read_exact(&mut data)?;
+                Ok(Some(data))
+            }
+            (None, Occupied(cached)) => {
+                debug!("getting from write cache");
+                Ok(Some(cached.get().clone()))
+            }
+            (None, Vacant(_)) => {
+                debug!("not found");
+                Ok(None)
+            }
+        }
+    }
+
+    /// # Panics
+    /// See [`Self::new`].
+    #[tracing::instrument(skip(self, block))]
+    fn put_keyed(&self, k: &Cid, block: &[u8]) -> anyhow::Result<()> {
+        match self.write_cache.lock().entry(*k) {
+            Occupied(already) if already.get() == block => {
+                debug!("already in cache");
+                Ok(())
+            }
+            Occupied(_) => panic!("mismatched data for cid {}", k),
+            Vacant(vacant) => {
+                debug!(bytes = block.len(), "insert into cache");
+                vacant.insert(block.to_owned());
+                Ok(())
+            }
+        }
+    }
+}
+
+fn read_header(reader: &mut impl BufRead) -> io::Result<CarHeader> {
+    let header_len = read_usize(reader)?;
+    match reader.fill_buf()? {
+        buf if buf.is_empty() => Err(io::Error::from(UnexpectedEof)),
+        nonempty if nonempty.len() < header_len => Err(io::Error::new(
+            UnexpectedEof,
+            "header is too short, or BufReader doesn't have enough capacity for a header",
+        )),
+        header_etc => match fvm_ipld_encoding::from_slice(&header_etc[..header_len]) {
+            Ok(header) => {
+                reader.consume(header_len);
+                Ok(header)
+            }
+            Err(e) => Err(io::Error::new(InvalidData, e)),
+        },
+    }
+}
+
+// Importantly, we seek _past_ the data, rather than read any in.
+// This allows us to keep indexing fast.
+fn read_block(reader: &mut (impl BufRead + Seek)) -> Option<cid::Result<(Cid, BlockLocation)>> {
+    match reader.fill_buf() {
+        Ok(buf) if buf.is_empty() => None, // EOF
+        Ok(_nonempty) => match (
+            read_usize(reader),
+            reader.stream_position(),
+            Cid::read_bytes(&mut *reader),
+        ) {
+            (Ok(length), Ok(offset), Ok(cid)) => {
+                let next_block_offset = offset + u64::try_from(length).unwrap();
+                if let Err(e) = reader.seek(SeekFrom::Start(next_block_offset)) {
+                    return Some(Err(cid::Error::Io(e)));
+                }
+                Some(Ok((cid, BlockLocation { offset, length })))
+            }
+            (Err(e), _, _) | (_, Err(e), _) => Some(Err(cid::Error::Io(e))),
```

Comment:
nit: Perhaps some simple comments for the error cases here could also be beneficial as it gets a little convoluted with all the different types of errors and nesting.